File: app/src/main/java/com/sunny/lib/image/FrescoUtils.java
package com.sunny.lib.image;

import android.content.Context;
import android.graphics.Bitmap;
import android.net.Uri;
import android.text.TextUtils;

import com.facebook.cache.disk.DiskCacheConfig;
import com.facebook.common.executors.UiThreadImmediateExecutorService;
import com.facebook.common.internal.Supplier;
import com.facebook.common.references.CloseableReference;
import com.facebook.common.util.ByteConstants;
import com.facebook.datasource.DataSource;
import com.facebook.drawee.backends.pipeline.Fresco;
import com.facebook.drawee.interfaces.DraweeController;
import com.facebook.drawee.view.SimpleDraweeView;
import com.facebook.imagepipeline.cache.MemoryCacheParams;
import com.facebook.imagepipeline.common.ResizeOptions;
import com.facebook.imagepipeline.core.ImagePipeline;
import com.facebook.imagepipeline.core.ImagePipelineConfig;
import com.facebook.imagepipeline.core.ImagePipelineFactory;
import com.facebook.imagepipeline.datasource.BaseBitmapDataSubscriber;
import com.facebook.imagepipeline.image.CloseableImage;
import com.facebook.imagepipeline.request.ImageRequest;
import com.facebook.imagepipeline.request.ImageRequestBuilder;
import com.sunny.family.R;
import com.sunny.lib.utils.AppConfigUtils;
import com.sunny.lib.utils.HandlerUtils;
import com.sunny.lib.utils.SunLog;

import org.jetbrains.annotations.Nullable;

/**
 * author: wenhao
 * created on: 2019-05-29 16:50
 * description:
 */
public class FrescoUtils {
    private static final String TAG = "FrescoUtils";
    private static int count = 0;

    public static void init(Context context) {
        SunLog.INSTANCE.i("FrescoUtils", "init");

        DiskCacheConfig diskCacheConfig = DiskCacheConfig.newBuilder(context)
                .setMaxCacheSize(40 * ByteConstants.MB) // 最大磁盘空间
                .build();

        final MemoryCacheParams bitmapCacheParams = new MemoryCacheParams(40 * ByteConstants.MB, // 内存缓存中总图片的最大大小,以字节为单位。
                256, // 内存缓存中图片的最大数量。
                10 * ByteConstants.MB, // 内存缓存中准备清除但尚未被删除的总图片的最大大小,以字节为单位。
                256, // 内存缓存中准备清除的总图片的最大数量。
                2 * ByteConstants.MB);// 内存缓存中单个图片的最大大小。

        Supplier<MemoryCacheParams> mSupplierMemoryCacheParams = new Supplier<MemoryCacheParams>() {
            @Override
            public MemoryCacheParams get() {
                return bitmapCacheParams;
            }
        };

        ImagePipelineConfig config = ImagePipelineConfig.newBuilder(context)
                .setDownsampleEnabled(true).setResizeAndRotateEnabledForNetwork(true)
                .setMainDiskCacheConfig(diskCacheConfig)
                .setBitmapMemoryCacheParamsSupplier(mSupplierMemoryCacheParams)
                .setBitmapsConfig(AppConfigUtils.INSTANCE.isLowCostDevice()
                        ? Bitmap.Config.RGB_565
                        : null)
                .build();
        Fresco.initialize(context, config);
    }

    public static void setImageURI(String url, final SimpleDraweeView simpleDraweeView) {
        SunLog.INSTANCE.i(TAG, "setImageURI  count :" + count++);

        if (simpleDraweeView == null) {
            return;
        }

        if (TextUtils.isEmpty(url)) {
            return;
        }

        final Uri uri = (url != null) ? Uri.parse(url) : null;

        HandlerUtils.INSTANCE.getUiHandler().post(new Runnable() {
            @Override
            public void run() {
                DraweeController controller = Fresco.newDraweeControllerBuilder().setUri(uri)
                        .setAutoPlayAnimations(true)
                        .setOldController(simpleDraweeView.getController()).build();
                simpleDraweeView.setController(controller);
            }
        });
    }

    public static void setImageURI(String url, final SimpleDraweeView simpleDraweeView, int width,
                                   int height) {
        SunLog.INSTANCE.i(TAG, "setImageURI2  count :" + count++);

        if (simpleDraweeView == null) {
            return;
        }

        if (simpleDraweeView.getController() != null) {
            return;
        }

        if (TextUtils.isEmpty(url)) {
            return;
        }

        final Uri uri = (url != null) ? Uri.parse(url) : null;

        final ImageRequest request = ImageRequestBuilder.newBuilderWithSource(uri)
                .setResizeOptions(new ResizeOptions(width, height)).build();

        HandlerUtils.INSTANCE.getUiHandler().post(new Runnable() {
            @Override
            public void run() {
                DraweeController controller = Fresco.newDraweeControllerBuilder().setUri(uri)
                        .setAutoPlayAnimations(true)
                        .setOldController(simpleDraweeView.getController()).setImageRequest(request)
                        .build();

                simpleDraweeView.setController(controller);
            }
        });
    }

    public static void resetImageURI(String url, final SimpleDraweeView simpleDraweeView, int width,
                                     int height) {
        SunLog.INSTANCE.i(TAG, "setImageURI2  count :" + count++);
        if (simpleDraweeView == null) {
            return;
        }

        if (TextUtils.isEmpty(url)) {
            return;
        }

        Object tag = simpleDraweeView.getTag(R.id.tag_imageview_url);
        if (tag instanceof String && url.equals(tag)) {
            // Logger.i(TAG, "setImageURI2 tag :" + tag);
            return;
        } else {
            // Logger.i(TAG, "setImageURI2 url :" + url);
            simpleDraweeView.setTag(R.id.tag_imageview_url, url);
        }

        final Uri uri = (url != null) ? Uri.parse(url) : null;

        final ImageRequest request = ImageRequestBuilder.newBuilderWithSource(uri)
                .setResizeOptions(new ResizeOptions(width, height)).build();

        HandlerUtils.INSTANCE.getUiHandler().post(new Runnable() {
            @Override
            public void run() {
                DraweeController controller = Fresco.newDraweeControllerBuilder().setUri(uri)
                        .setAutoPlayAnimations(true)
                        .setOldController(simpleDraweeView.getController()).setImageRequest(request)
                        .build();

                simpleDraweeView.setController(controller);
            }
        });
    }

    public static void setImageURILow(String url, String lowUrl,
                                      final SimpleDraweeView simpleDraweeView, int width, int height) {
        if (simpleDraweeView == null) {
            return;
        }

        if (TextUtils.isEmpty(url)) {
            return;
        }


        final Uri uri = (url != null) ? Uri.parse(url) : null;

        final ImageRequest request = ImageRequestBuilder.newBuilderWithSource(uri)
                .setResizeOptions(new ResizeOptions(width, height)).build();

        ImageRequest lowImageReq = null;
        if (!TextUtils.isEmpty(lowUrl)) {
            lowImageReq = ImageRequestBuilder.newBuilderWithSource(Uri.parse(lowUrl))
                    .setResizeOptions(new ResizeOptions(width, height)).build();
        }

        final ImageRequest lowImageRequest = lowImageReq;

        HandlerUtils.INSTANCE.getUiHandler().post(new Runnable() {
            @Override
            public void run() {
                DraweeController controller = Fresco.newDraweeControllerBuilder().setUri(uri)
                        .setAutoPlayAnimations(true)
                        .setOldController(simpleDraweeView.getController()).setImageRequest(request)
                        .setLowResImageRequest(lowImageRequest).build();

                simpleDraweeView.setController(controller);
            }
        });
    }

    public static void setImageRes(int imageRes, SimpleDraweeView simpleDraweeView) {
        SunLog.INSTANCE.i(TAG, "setImageRes  count :" + count++);

        if (simpleDraweeView == null) {
            return;
        }
        // GenericDraweeHierarchy hierarchy = simpleDraweeView.getHierarchy();
        // hierarchy.setActualImageScaleType(ScalingUtils.ScaleType.FIT_CENTER);
        simpleDraweeView.setActualImageResource(imageRes);
    }

    /**
     * 继续Fresco图片加载，旧平台的Fresco版本不支持，该方法已兼容
     */
    public static void resume() {
//        if (!EuiUtils.isOldPlatform() || ModuleUtil.isModule()) {
        ImagePipeline imagePipeline = Fresco.getImagePipeline();
        if (imagePipeline.isPaused()) {
            imagePipeline.resume();
        }
//        }
    }

    /**
     * 暂停Fresco图片加载，旧平台的Fresco版本不支持，该方法已兼容
     */
    public static void pause() {
//        if (!SDKUtil.isOldPlatform() || ModuleUtil.isModule()) {
        ImagePipeline imagePipeline = Fresco.getImagePipeline();
        if (!imagePipeline.isPaused()) {
            imagePipeline.pause();
        }
//        }
    }

    /**
     * 清除图片缓存
     */
    public static void clearMemoryCache() {
        ImagePipeline imagePipeline = Fresco.getImagePipeline();
        imagePipeline.clearMemoryCaches();
        // imagePipeline.clearDiskCaches();
    }

    public static void clearCache() {
        ImagePipeline imagePipeline = Fresco.getImagePipeline();
        // combines above two lines
        imagePipeline.clearCaches();
    }

    /**
     * 清除单张图片的磁盘缓存
     */
    public static void clearCacheByUrl(String url) {
        ImagePipeline imagePipeline = Fresco.getImagePipeline();
        Uri uri = Uri.parse(url);
        // imagePipeline.evictFromMemoryCache(uri);
        // imagePipeline.evictFromDiskCache(uri);
        imagePipeline.evictFromCache(uri);// 这个包含了从内存移除和从硬盘移除
    }

    public static void getImage(String url, final FrescoImageLoadingListener imageLoadingListener) {
        if (TextUtils.isEmpty(url)) {
            return;
        }

        if (imageLoadingListener == null) {
            return;
        }

        final Uri uri = (url != null) ? Uri.parse(url) : null;

        ImageRequestBuilder requestBuilder = ImageRequestBuilder.newBuilderWithSource(uri);
        ImageRequest imageRequest = requestBuilder.build();
        DataSource<CloseableReference<CloseableImage>> dataSource = ImagePipelineFactory
                .getInstance().getImagePipeline().fetchDecodedImage(imageRequest, null);
        dataSource.subscribe(new BaseBitmapDataSubscriber() {
            @Override
            protected void onNewResultImpl(@Nullable Bitmap bitmap) {
                if (bitmap != null && !bitmap.isRecycled()) {
                    imageLoadingListener.onSuccess(uri, bitmap);
                }
            }

            @Override
            protected void onFailureImpl(
                    DataSource<CloseableReference<CloseableImage>> dataSource) {
                Throwable throwable = null;
                if (dataSource != null) {
                    throwable = dataSource.getFailureCause();
                }
                imageLoadingListener.onFailure(uri, throwable);
            }
        }, UiThreadImmediateExecutorService.getInstance());
    }
}

File: libCommon/src/main/java/com/sunny/lib/common/fresco/FrescoUtils.kt
package com.sunny.lib.common.fresco

import android.content.ComponentCallbacks2
import android.content.Context
import android.net.Uri
import android.os.Handler
import android.text.TextUtils
import com.facebook.common.memory.MemoryTrimType
import com.facebook.drawee.backends.pipeline.Fresco
import com.facebook.imagepipeline.common.ResizeOptions
import com.facebook.imagepipeline.request.ImageRequestBuilder
import com.sunny.lib.common.utils.DomainUtils
import com.sunny.lib.common.utils.HandlerUtils
import com.sunny.lib.base.log.SunLog
import com.sunny.lib.utils.AppConfigUtils

object FrescoUtils {

    const val TAG = "FrescoTag"

    private val frescoHelper by lazy {
        FrescoCacheHelper()
    }

    /**
     * 日志是否打开
     */
    @JvmStatic
    fun logEnable(): Boolean {
        return true
    }

    /**
     * 是否是低端设备
     */
    @JvmStatic
    fun lowDevice(): Boolean {
        return AppConfigUtils.isLowCostDevice()
    }


    @JvmStatic
    fun init(context: Context) {

        val config = frescoHelper.buildConfig(context)
        Fresco.initialize(context, config)
    }

    /**
     * 内存不足
     */
    @JvmStatic
    fun onLowMemory() {
        SunLog.i(TAG, "onLowMemory")
        frescoHelper.tryTrimMemoryCache(MemoryTrimType.OnAppBackgrounded)
    }

    /**
     * 内存不足时的通知，无论当前应用是前台还是后台，
     * 只要内存不足，就进行图片缓存的释放
     */
    @JvmStatic
    fun onTrimMemory(level: Int) {
        SunLog.i(TAG, "onTrimMemory  $level")

        val type = when (level) {

            ComponentCallbacks2.TRIM_MEMORY_BACKGROUND,
            ComponentCallbacks2.TRIM_MEMORY_COMPLETE,
            ComponentCallbacks2.TRIM_MEMORY_MODERATE ->
                MemoryTrimType.OnSystemLowMemoryWhileAppInBackground /* 内存不足，当前应用在后台 */

            ComponentCallbacks2.TRIM_MEMORY_RUNNING_CRITICAL,
            ComponentCallbacks2.TRIM_MEMORY_RUNNING_LOW ->
                MemoryTrimType.OnSystemLowMemoryWhileAppInForeground /* 内存不足，当前应用在前台 */

//            ComponentCallbacks2.TRIM_MEMORY_RUNNING_MODERATE ->
//                MemoryTrimType.OnSystemMemoryCriticallyLowWhileAppInForeground /* 内存严重不足，当前应用在前台 */

            else -> MemoryTrimType.OnAppBackgrounded /* 应用进入后台 */
        }

        frescoHelper.tryTrimMemoryCache(type)
    }

    /**
     * 继续Fresco图片加载，旧平台的Fresco版本不支持，该方法已兼容
     */
    @JvmStatic
    fun resume() {
        if (lowDevice()) {
            return
        }

        val pipeline = Fresco.getImagePipeline()
        if (pipeline.isPaused) {
            pipeline.resume()
        }
    }

    /**
     * 暂停Fresco图片加载，旧平台的Fresco版本不支持，该方法已兼容
     */
    @JvmStatic
    fun pause() {
        if (lowDevice()) {
            return
        }

        val pipeline = Fresco.getImagePipeline()
        if (!pipeline.isPaused) {
            pipeline.pause()
        }

    }


    private fun buildUri(url: String?): Uri? {
        if (TextUtils.isEmpty(url)) {
            return null
        }

        val domainUrl = DomainUtils.urlDomainReplace(url)

        return Uri.parse(domainUrl)
    }

    private fun getUIHandler(): Handler {
        return HandlerUtils.uiHandler
    }

    @JvmStatic
    fun loadImageUrl(url: String?, view: LeFrescoImageView?) {
        if (view == null) {
            return
        }

        if (view.controller != null) {
            return
        }

        val uri: Uri = buildUri(url) ?: return

        getUIHandler().post {
            view.controller = Fresco.newDraweeControllerBuilder()
                    .setUri(uri)
                    .setAutoPlayAnimations(true)
                    .setOldController(view.controller)
                    .build()
        }
    }

    @JvmStatic
    fun loadImageUrl(url: String?, view: LeFrescoImageView?, width: Int, height: Int) {
        if (view == null) {
            return
        }

        val uri: Uri = buildUri(url) ?: return

        val request = ImageRequestBuilder.newBuilderWithSource(uri)
                .setResizeOptions(ResizeOptions(width, height)).build()

        getUIHandler().post {
            view.controller = Fresco.newDraweeControllerBuilder()
                    .setUri(uri)
                    .setAutoPlayAnimations(true)
                    .setOldController(view.controller)
                    .setImageRequest(request)
                    .build()
        }
    }

}
